### PR TITLE
Add rule for check calling method in object method

### PR DESCRIPTION
**Pull request by romanmatyus**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # Change Log
 
 ## [Unreleased][unreleased]
+### Added
+- Rule for checking call `setRequired()` in method `globalConfiguration` - for Notcms plugins
```
Comment (lulco) [inline CHANGELOG.md]: Should be pageConfiguration, not globalConfiguration

File: src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (get_class($node) !== MethodCall::class) {
+            return [];
+        }
+        if ($node->name->name !== 'setRequired') {
+            return [];
+        }
+        if ($scope->getFunctionName() !== 'globalConfiguration') {
```
Comment (lulco) [inline src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php]: Page

File: src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        return MethodCall::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (get_class($node) !== MethodCall::class) {
+            return [];
+        }
+        if ($node->name->name !== 'setRequired') {
```
Comment (lulco) [inline src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php]: We need to test caller too. There can be setRequire() call also on non-ConfigItems

File: src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+final class PluginConfigItemNoRequiredInGlobalConfigurationRule implements Rule
+{
+    public function getNodeType(): string
+    {
+        return MethodCall::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (get_class($node) !== MethodCall::class) {
```
Comment (lulco) [inline src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php]: Instanceof would be better

File: src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            return [];
+        }
+        if ($node->name->name !== 'setRequired') {
+            return [];
+        }
+        if ($scope->getFunctionName() !== 'globalConfiguration') {
+            return [];
+        }
+        if (!$scope->isInClass()) {
```
Comment (lulco) [inline src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php]: Maybe we should check if it is in class of type PluginDefinition

**Comment by lulco** [comment]
Is it working without some source files of notcms/core? Some stubs

**Review by lulco — CHANGES_REQUESTED** [review]

**Comment by romanmatyus** [comment]
@lulco Can I add some stub files from Notcms? License etc..?

**Comment by lulco** [comment]
> @lulco Can I add some stub files from Notcms? License etc..?

It would be better to create rule more generic - in class, in method, caller class and called method name should be in parameters and the rule would be registered in final config

**Comment by lulco** [comment]
Also, please add new rule to readme

File: src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php (hunk 5 above)
Comment (romanmatyus) [reply src/Rule/Notcms/PluginConfigItemNoRequiredInGlobalConfigurationRule.php]: It's possible only with load real package. Stub files not support reflection for this.

File: CHANGELOG.md (hunk 1 above)
Comment (lulco) [inline CHANGELOG.md]: rule for check calling method in object method

File: README.md [hunk 6]
```diff
@@ -138,3 +138,45 @@ final class SomeClass implements MyInterface
 }
 ```
 :+1:
+
+### Check calling method in object method
+Checks if some method is not used in disabled context - specific method of object.
+```neon
+services:
+    disableCallMethodInObjectMethodRule:
+        factory: Efabrica\PHPStanRules\Rule\General\DisableCallMethodInObjectMethodRule('pageConfiguration', 'setRequired')
```
Comment (lulco) [inline README.md]: I miss some other context: context class, context method, caller class, called method. You have just context method and called method. I am not sure if it is enough

File: composer.json [hunk 7]
```diff
@@ -7,6 +7,7 @@
         "phpstan/phpstan": "^1.6"
     },
     "require-dev": {
+        "tracy/tracy": "*",
```
Comment (lulco) [inline composer.json:10]: we don't need tracy here

File: rules.neon [hunk 8]
```diff
@@ -1,3 +1,8 @@
+parameters:
+	stubFiles:
```
Comment (lulco) [inline rules.neon]: use spaces

File: src/Rule/General/DisableCallMethodInObjectMethodRule.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        if (
+            !$node instanceof MethodCall ||
+            $node->name->name !== $this->calledMethod ||
+            $scope->getFunctionName() !== $this->objectMethod ||
+            !$scope->isInClass()
+        ) {
+            return [];
+        }
+        $file = $scope->getFile();
+        $errors[] = RuleErrorBuilder::message('Method ' . $scope->getClassReflection()->getName() . '::' . $scope->getFunctionName() . ' is called with ' . $node->name->name . '() option.')->file($file)->line($node->getLine())->build();
```
Comment (lulco) [inline src/Rule/General/DisableCallMethodInObjectMethodRule.php]: is called with option is little bit wrong, you should tell user that he uses method in context of class / method

File: src/stubs/BasePluginDefinition.stub [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+use Efabrica\Cms\Core\Plugin\Config\Factory\ConfigItemFactoryStorage;
+use InvalidArgumentException;
+use Nette\Application\UI\Control;
+use Nette\ComponentModel\IContainer;
+use Nette\Database\Table\IRow;
+use Nette\DI\Container;
+use Nette\Localization\ITranslator;
+use ReflectionClass;
+
+abstract class BasePluginDefinition implements PluginDefinitionInterface
```
Comment (lulco) [inline src/stubs/BasePluginDefinition.stub]: no need to use whole class, just some sampe methods

File: tests/Rule/General/DisableCallMethodInObjectMethodRule/Fixtures/PluginWithSetRequired.php [hunk 11]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace BaseModule\Plugin\Header;
+
+class PluginWithSetRequired extends BasePluginDefinition
+{
+    public function anotherMethod(): array
+    {
+        return [
+            new ChoozerConfigItem('main_menu_parent_page_id', 'Stránka s hlavným menu', 'page'),
```
Comment (lulco) [inline tests/Rule/General/DisableCallMethodInObjectMethodRule/Fixtures/PluginWithSetRequired.php]: english please

File: src/stubs/BasePluginDefinition.stub (hunk 10 above)
Comment (mabar) [reply src/stubs/BasePluginDefinition.stub]: Oh man, don't you have headaches from such classes? :'D

Comment (lulco) [reply src/stubs/BasePluginDefinition.stub]: Not at all :)

**Comment by lulco** [comment]
Thanks I'll finish some cleanup later

**Pull request merged**